### ISSUE TITLE
[Memory leak] rocksdb.block-cache-size has no effect in single mode

### ISSUE THREAD

**Issue by daoxian** [arangodb/arangodb#9298]
## My Environment

* __ArangoDB Version__:        3.4.5
* __Storage Engine__:              RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   ArangoDB Starter-->
* __Operating System__:        CentOS7.4
* __Total RAM in your machine__:       54G
* __Disks in use__:        10%
* __Used Package__:             RedHat .rpm

After I started the command:
` arangodb --starter.mode=single --server.storage-engine=rocksdb --starter.data-dir=/home/admin/arangodb/data_3.4.5_singlemode_partition --all.rocksdb.block-cache-size=5000000000 --all.rocksdb.enforce-block-cache-size-limit=true`

the child process was created:
`/usr/sbin/arangod -c /home/admin/arangodb/data_3.4.5_singlemode_partition/single8529/arangod.conf --database.directory /home/admin/arangodb/data_3.4.5_singlemode_partition/single8529/data --javascript.startup-directory /usr/share/arangodb3/js --javascript.app-path /home/admin/arangodb/data_3.4.5_singlemode_partition/single8529/apps --log.file /home/admin/arangodb/data_3.4.5_singlemode_partition/single8529/arangod.log --log.force-direct false --javascript.copy-installation true --foxx.queues true --server.statistics true --rocksdb.block-cache-size 5000000000 --rocksdb.enforce-block-cache-size-limit true `

Then kept inserting and querying. Hours later, the memory consumption kept rising and finally larger than 54G and OOM killed.

 --rocksdb.enforce-block-cache-size-limit option seems effective in cluster mode, **but seems not work in single mode.**

**Comment by internalfx** [comment]
Use this flag instead...

`--all.rocksdb.block-cache-size=5000000000`

**Comment by internalfx** [comment]
```
arangodb \
  --server.storage-engine=rocksdb \
  --starter.mode=single \
  --all.rocksdb.block-cache-size=100000000
```

This is my start script and I can confirm that it is respecting the block cache size.

I'm on version `3.4.6-1` Maybe you can give the newer version a try.

**Comment by daoxian** [comment]
I've tried the new version 3.4.6-1, yield the same OOM result.

![image](https://user-images.githubusercontent.com/34263430/59812029-f6d63680-933e-11e9-918c-05d1b21eec76.png)

**Comment by daoxian** [comment]
Any suggestions? @jsteemann @graetzer

**Comment by graetzer** [comment]
Are you running the processes on different servers

**Comment by daoxian** [comment]
> Are you running the processes on different servers ?

No, arangodb and arangod processes are on the same server.

**Comment by graetzer** [comment]
The limit applies to each cluster process individually, if you start five servers you should lower the limit

**Comment by daoxian** [comment]
> The limit applies to each cluster process individually, if you start five servers you should lower the limit
> […](#)

Only 1 arangod process running on each machine.  Every machine's memory consumption keeps rising until OOM.  This is very likely a bug.

**Comment by graetzer** [comment]
Did you set the total write buffer size ?

**Comment by daoxian** [comment]
> Did you set the total write buffer size ?
> […](#)
> Am 20.06.2019 um 04:05 schrieb daoxian ***@***.***>: The limit applies to each cluster process individually, if you start five servers you should lower the limit … Only 1 arangod process running on each machine. Every machine's memory consumption keeps rising until OOM. This is very likely a bug. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or mute the thread.

I'm now using 3.4.6-1, with --rocksdb.total-write-buffer-size options.  But memory consumption is still rising.

`115357 ?        S      0:00 sudo nohup arangodb --starter.mode=single --server.storage-engine=rocksdb --starter.data-dir=/home/admin/arangodb/data_3.4.5_singlemode_partition --all.rocksdb.block-cache-size=5000000000 --all.rocksdb.enforce-block-cache-size-limit=true --all.rocksdb.total-write-buffer-size=500000000
115371 ?        Sl     0:00  \_ arangodb --starter.mode=single --server.storage-engine=rocksdb --starter.data-dir=/home/admin/arangodb/data_3.4.5_singlemode_partition --all.rocksdb.block-cache-size=5000000000 --all.rocksdb.enforce-block-cache-size-limit=true --all.rocksdb.total-write-buffer-size=500000000
115395 ?        Sl   12204:09      \_ /usr/sbin/arangod -c /home/admin/arangodb/data_3.4.5_singlemode_partition/single8529/arangod.conf --database.directory /home/admin/arangodb/data_3.4.5_singlemode_partition/single8529/data --javascript.startup-directory /usr/share/arangodb3/js --javascript.app-path /home/admin/arangodb/data_3.4.5_singlemode_partition/single8529/apps --log.file /home/admin/arangodb/data_3.4.5_singlemode_partition/single8529/arangod.log --log.force-direct false --javascript.copy-installation true --foxx.queues true --server.statistics true --rocksdb.enforce-block-cache-size-limit true --rocksdb.total-write-buffer-size 500000000 --rocksdb.block-cache-size 5000000000
`

![image](https://user-images.githubusercontent.com/34263430/59840431-e779dc00-9384-11e9-821e-6854f3a49641.png)

**Comment by daoxian** [comment]
Could anyone please help me?  I've been blocked by this problem for several days...

**Comment by daoxian** [comment]
I've found one of the possible reason:  using async API, 'return_result' is set to True, which caused the server side to save job info and never release.

**Comment by graetzer** [comment]
Ah yes, I think there might not be a ttl for these results

**Comment by internalfx** [comment]
@graetzer Oh Right! ....I vaguely remembering there being a case where if the cursor is not fully read, it never closes. I think you're on to something.

**Comment by daoxian** [comment]
Should collections get unloaded after used?  I observed that after bulk import docs into a collection, then unloaded the collection from memory,  the memory consumption is recovered steadily. I'm not sure if it's the real case.

**Comment by daoxian** [comment]
@graetzer   It's a surprise to me that memory consumption doesn't respect the max limit startup option if the size of documents are beyond the limi. And more imortant, each worker thread consumes quantity of memory when parallel large query or large insertion happens.

**Comment by frankcb298** [comment]
hi @daoxian have you fixed this problem yet?

**Comment by graetzer** [comment]
@graetzer can you elaborate on this ? Total-write buffer size should be sufficient to limit rocksdb's ram usage on inserts. If you are referring to queries you might need to adjust the memory limit for AQL

**Comment by daoxian** [comment]
> hi @daoxian have you fixed this memory problem yet?

Thanks for you care.  Not yet,   I've worked around this problem by limiting a far smaller value than the server's total memory.  The problem is all the limitations are not global scope, but per-thread or per-request.

**Comment by dothebart** [comment]
As discussed in https://github.com/arangodb/arangodb/issues/11577 [exposing more options of rocksdb to the commandline](https://www.arangodb.com/docs/stable/programs-arangod-options.html#rocksdb) seems to ease the memory situation.

Namely:
 -   `--rocksdb.cache-index-and-filter-blocks-with-high-priority`
 -   `--rocksdb.pin-l0-filter-and-index-blocks-in-cache`
 -   `--rocksdb.pin-top-level-index-and-filter`

These have been added to [the now released ArangoDB 3.7](https://www.arangodb.com/2020/08/arangodb-3-7-a-big-step-forward-for-multi-model/) - hence closing this.
